File: stick-rumble-client/src/shared/weaponConfig.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import {
  loadWeaponConfigs,
  getWeaponConfig,
  getWeaponConfigSync,
  validateWeaponConfig,
  parseHexColor,
  resetWeaponConfigs,
  type WeaponConfig,
} from './weaponConfig';

describe('weaponConfig', () => {
  describe('parseHexColor', () => {
    it('should parse hex color string with 0x prefix', () => {
      expect(parseHexColor('0xffaa00')).toBe(0xffaa00);
      expect(parseHexColor('0xFF0000')).toBe(0xff0000);
      expect(parseHexColor('0x00FF00')).toBe(0x00ff00);
    });

    it('should parse hex color string without 0x prefix', () => {
      expect(parseHexColor('ffaa00')).toBe(0xffaa00);
      expect(parseHexColor('FF0000')).toBe(0xff0000);
    });
  });

  describe('validateWeaponConfig', () => {
    it('should validate a valid weapon config', () => {
      const config: WeaponConfig = {
        name: 'Pistol',
        damage: 25,
        fireRate: 3.0,
        magazineSize: 15,
        reloadTimeMs: 1500,
        projectileSpeed: 800,
        range: 800,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: null,
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors).toHaveLength(0);
    });

    it('should detect missing name', () => {
      const config: WeaponConfig = {
        name: '',
        damage: 25,
        fireRate: 3.0,
        magazineSize: 15,
        reloadTimeMs: 1500,
        projectileSpeed: 800,
        range: 800,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: null,
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors).toContain('Weapon name cannot be empty');
    });

    it('should detect invalid damage', () => {
      const config: WeaponConfig = {
        name: 'Test',
        damage: 0,
        fireRate: 3.0,
        magazineSize: 15,
        reloadTimeMs: 1500,
        projectileSpeed: 800,
        range: 800,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: null,
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors.some((e) => e.includes('damage must be positive'))).toBe(true);
    });

    it('should detect invalid fire rate', () => {
      const config: WeaponConfig = {
        name: 'Test',
        damage: 25,
        fireRate: 0,
        magazineSize: 15,
        reloadTimeMs: 1500,
        projectileSpeed: 800,
        range: 800,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: null,
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors.some((e) => e.includes('fire rate must be positive'))).toBe(true);
    });

    it('should detect invalid range', () => {
      const config: WeaponConfig = {
        name: 'Test',
        damage: 25,
        fireRate: 3.0,
        magazineSize: 15,
        reloadTimeMs: 1500,
        projectileSpeed: 800,
        range: 0,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: null,
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors.some((e) => e.includes('range must be positive'))).toBe(true);
    });

    it('should validate ranged weapon has projectile speed', () => {
      const config: WeaponConfig = {
        name: 'InvalidGun',
        damage: 25,
        fireRate: 3.0,
        magazineSize: 15, // Has ammo but no projectile speed
        reloadTimeMs: 1500,
        projectileSpeed: 0,
        range: 800,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: null,
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors.some((e) => e.includes('projectile speed'))).toBe(true);
    });

    it('should validate recoil parameters if present', () => {
      const config: WeaponConfig = {
        name: 'Test',
        damage: 25,
        fireRate: 3.0,
        magazineSize: 15,
        reloadTimeMs: 1500,
        projectileSpeed: 800,
        range: 800,
        arcDegrees: 0,
        knockbackDistance: 0,
        recoil: {
          verticalPerShot: 2.0,
          horizontalPerShot: 0,
          recoveryTime: 0, // Invalid
          maxAccumulation: -5, // Invalid
        },
        spreadDegrees: 0,
        visuals: {
          muzzleFlashColor: '0xffdd00',
          muzzleFlashSize: 8,
          muzzleFlashDuration: 50,
        },
      };

      const errors = validateWeaponConfig(config);
      expect(errors.some((e) => e.includes('recovery time'))).toBe(true);
      expect(errors.some((e) => e.includes('max accumulation'))).toBe(true);
    });
  });

  describe('getWeaponConfigSync', () => {
    it('should return hardcoded config when not loaded', () => {
      const pistol = getWeaponConfigSync('Pistol');
      expect(pistol).not.toBeNull();
      expect(pistol?.name).toBe('Pistol');
      expect(pistol?.damage).toBe(25);
    });

    it('should return null for invalid weapon', () => {
      const invalid = getWeaponConfigSync('InvalidWeapon');
      expect(invalid).toBeNull();
    });

    it('should return configs for all weapons', () => {
      const weapons = ['Pistol', 'Bat', 'Katana', 'Uzi', 'AK47', 'Shotgun'];

      for (const weaponName of weapons) {
        const config = getWeaponConfigSync(weaponName);
        expect(config).not.toBeNull();
        expect(config?.name).toBe(weaponName);
      }
    });
  });

  describe('getWeaponConfig', () => {
    it('should warn and return hardcoded config when configs not loaded', () => {
      const consoleWarnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
      resetWeaponConfigs();

      const pistol = getWeaponConfig('Pistol');
      expect(pistol).not.toBeNull();
      expect(pistol?.name).toBe('Pistol');
      expect(consoleWarnSpy).toHaveBeenCalledWith(
        'Weapon configs not loaded yet, call loadWeaponConfigs() first'
      );

      consoleWarnSpy.mockRestore();
    });
  });

  describe('loadWeaponConfigs', () => {
    beforeEach(() => {
      // Reset global state
      vi.clearAllMocks();
      // Reset the weapon configs cache
      resetWeaponConfigs();
    });

    it('should return cached configs on second call', async () => {
      const mockConfigs = {
        version: '1.0.0',
        weapons: {
          TestWeapon: {
            name: 'TestWeapon',
            damage: 50,
            fireRate: 5.0,
            magazineSize: 20,
            reloadTimeMs: 2000,
            projectileSpeed: 900,
            range: 700,
            arcDegrees: 0,
            knockbackDistance: 0,
            recoil: null,
            spreadDegrees: 0,
            visuals: {
              muzzleFlashColor: '0xff0000',
              muzzleFlashSize: 10,
              muzzleFlashDuration: 60,
            },
          },
        },
      };

      let fetchCallCount = 0;
      global.fetch = vi.fn().mockImplementation(() => {
        fetchCallCount++;
        return Promise.resolve({
          ok: true,
          json: async () => mockConfigs,
        } as Response);
      });

      // First call should fetch
      const configs1 = await loadWeaponConfigs();
      expect(fetchCallCount).toBe(1);
      expect(configs1.TestWeapon).toBeDefined();

      // Second call should return cached (no new fetch)
      const configs2 = await loadWeaponConfigs();
      expect(fetchCallCount).toBe(1); // Still 1, not 2
      expect(configs2.TestWeapon).toBeDefined();
      expect(configs2).toBe(configs1); // Same reference
    });

    it('should load configs from fetch', async () => {
      const mockConfigs = {
        version: '1.0.0',
        weapons: {
          TestWeapon: {
            name: 'TestWeapon',
            damage: 50,
            fireRate: 5.0,
            magazineSize: 20,
            reloadTimeMs: 2000,
            projectileSpeed: 900,
            range: 700,
            arcDegrees: 0,
            knockbackDistance: 0,
            recoil: null,
            spreadDegrees: 0,
            visuals: {
              muzzleFlashColor: '0xff0000',
              muzzleFlashSize: 10,
              muzzleFlashDuration: 60,
            },
          },
        },
      };

      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => mockConfigs,
      } as Response);

      const configs = await loadWeaponConfigs();
      expect(configs).toBeDefined();
      expect(configs.TestWeapon).toBeDefined();
      expect(configs.TestWeapon.name).toBe('TestWeapon');
      expect(configs.TestWeapon.damage).toBe(50);
    });

    it('should fallback to hardcoded configs on fetch error', async () => {
      global.fetch = vi.fn().mockRejectedValue(new Error('Network error'));

      const configs = await loadWeaponConfigs();
      expect(configs).toBeDefined();
      expect(configs.Pistol).toBeDefined();
      expect(configs.Pistol.name).toBe('Pistol');
    });

    it('should fallback to hardcoded configs on non-ok response', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: false,
        statusText: 'Not Found',
      } as Response);

      const configs = await loadWeaponConfigs();
      expect(configs).toBeDefined();
      expect(configs.Pistol).toBeDefined();
    });
  });

  describe('Weapon configurations', () => {
    it('should have valid Pistol config', () => {
      const pistol = getWeaponConfigSync('Pistol');
      expect(pistol).not.toBeNull();
      expect(pistol?.damage).toBe(25);
      expect(pistol?.fireRate).toBe(3.0);
      expect(pistol?.magazineSize).toBe(15);
      expect(pistol?.reloadTimeMs).toBe(1500);
      expect(pistol?.range).toBe(800);
    });

    it('should have valid Bat config', () => {
      const bat = getWeaponConfigSync('Bat');
      expect(bat).not.toBeNull();
      expect(bat?.damage).toBe(25);
      expect(bat?.fireRate).toBe(2.0);
      expect(bat?.magazineSize).toBe(0); // Melee
      expect(bat?.range).toBe(64);
      expect(bat?.arcDegrees).toBe(90);
      expect(bat?.knockbackDistance).toBe(40);
    });

    it('should have valid Katana config', () => {
      const katana = getWeaponConfigSync('Katana');
      expect(katana).not.toBeNull();
      expect(katana?.damage).toBe(45);
      expect(katana?.fireRate).toBe(1.25);
      expect(katana?.magazineSize).toBe(0); // Melee
      expect(katana?.range).toBe(80);
      expect(katana?.arcDegrees).toBe(90);
    });

    it('should have valid Uzi config', () => {
      const uzi = getWeaponConfigSync('Uzi');
      expect(uzi).not.toBeNull();
      expect(uzi?.damage).toBe(8);
      expect(uzi?.fireRate).toBe(10.0);
      expect(uzi?.magazineSize).toBe(30);
      expect(uzi?.reloadTimeMs).toBe(1500);
      expect(uzi?.range).toBe(600);
      expect(uzi?.recoil).not.toBeNull();
      expect(uzi?.recoil?.verticalPerShot).toBe(2.0);
      expect(uzi?.spreadDegrees).toBe(5.0);
    });

    it('should have valid AK47 config', () => {
      const ak47 = getWeaponConfigSync('AK47');
      expect(ak47).not.toBeNull();
      expect(ak47?.damage).toBe(20);
      expect(ak47?.fireRate).toBe(6.0);
      expect(ak47?.magazineSize).toBe(30);
      expect(ak47?.reloadTimeMs).toBe(2000);
      expect(ak47?.range).toBe(800);
      expect(ak47?.recoil).not.toBeNull();
    });

    it('should have valid Shotgun config', () => {
      const shotgun = getWeaponConfigSync('Shotgun');
      expect(shotgun).not.toBeNull();
      expect(shotgun?.damage).toBe(60);
      expect(shotgun?.fireRate).toBe(1.0);
      expect(shotgun?.magazineSize).toBe(6);
      expect(shotgun?.reloadTimeMs).toBe(2500);
      expect(shotgun?.range).toBe(300);
      expect(shotgun?.arcDegrees).toBe(15.0);
    });
  });
});

File: stick-rumble-server/internal/game/weapon_config_test.go
package game

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

func TestLoadWeaponConfigs_Success(t *testing.T) {
	// Create temporary config file
	tmpDir := t.TempDir()
	configPath := filepath.Join(tmpDir, "weapon-configs.json")

	config := WeaponConfigFile{
		Version: "1.0.0",
		Weapons: map[string]WeaponConfig{
			"Pistol": {
				Name:              "Pistol",
				Damage:            25,
				FireRate:          3.0,
				MagazineSize:      15,
				ReloadTimeMs:      1500,
				ProjectileSpeed:   800.0,
				Range:             800.0,
				ArcDegrees:        0,
				KnockbackDistance: 0,
				Recoil:            nil,
				SpreadDegrees:     0,
				Visuals: WeaponVisuals{
					MuzzleFlashColor:    "0xffdd00",
					MuzzleFlashSize:     8,
					MuzzleFlashDuration: 50,
				},
			},
		},
	}

	data, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		t.Fatalf("Failed to marshal config: %v", err)
	}

	if err := os.WriteFile(configPath, data, 0644); err != nil {
		t.Fatalf("Failed to write config file: %v", err)
	}

	// Test loading
	configs, err := LoadWeaponConfigs(configPath)
	if err != nil {
		t.Fatalf("LoadWeaponConfigs failed: %v", err)
	}

	if configs == nil {
		t.Fatal("Expected configs to be non-nil")
	}

	if configs["Pistol"] == nil {
		t.Fatal("Expected Pistol config to exist")
	}

	pistol := configs["Pistol"]
	if pistol.Name != "Pistol" {
		t.Errorf("Expected name 'Pistol', got '%s'", pistol.Name)
	}
	if pistol.Damage != 25 {
		t.Errorf("Expected damage 25, got %d", pistol.Damage)
	}
	if pistol.FireRate != 3.0 {
		t.Errorf("Expected fire rate 3.0, got %f", pistol.FireRate)
	}
}

func TestLoadWeaponConfigs_FileNotFound(t *testing.T) {
	_, err := LoadWeaponConfigs("/nonexistent/path/weapon-configs.json")
	if err == nil {
		t.Fatal("Expected error for nonexistent file, got nil")
	}
}

func TestLoadWeaponConfigs_InvalidJSON(t *testing.T) {
	tmpDir := t.TempDir()
	configPath := filepath.Join(tmpDir, "weapon-configs.json")

	// Write invalid JSON
	if err := os.WriteFile(configPath, []byte("invalid json {"), 0644); err != nil {
		t.Fatalf("Failed to write config file: %v", err)
	}

	_, err := LoadWeaponConfigs(configPath)
	if err == nil {
		t.Fatal("Expected error for invalid JSON, got nil")
	}
}

func TestLoadWeaponConfigs_AllWeapons(t *testing.T) {
	// Test with actual production config
	// Path from stick-rumble-server/internal/game/ to project root
	configPath := filepath.Join("..", "..", "..", "weapon-configs.json")

	configs, err := LoadWeaponConfigs(configPath)
	if err != nil {
		t.Fatalf("LoadWeaponConfigs failed: %v", err)
	}

	expectedWeapons := []string{"Pistol", "Bat", "Katana", "Uzi", "AK47", "Shotgun"}
	for _, weaponName := range expectedWeapons {
		if configs[weaponName] == nil {
			t.Errorf("Expected weapon '%s' to be in configs", weaponName)
		}
	}
}

func TestWeaponConfigToWeapon_Pistol(t *testing.T) {
	config := &WeaponConfig{
		Name:              "Pistol",
		Damage:            25,
		FireRate:          3.0,
		MagazineSize:      15,
		ReloadTimeMs:      1500,
		ProjectileSpeed:   800.0,
		Range:             800.0,
		ArcDegrees:        0,
		KnockbackDistance: 0,
		Recoil:            nil,
		SpreadDegrees:     0,
	}

	weapon := config.ToWeapon()

	if weapon.Name != "Pistol" {
		t.Errorf("Expected name 'Pistol', got '%s'", weapon.Name)
	}
	if weapon.Damage != 25 {
		t.Errorf("Expected damage 25, got %d", weapon.Damage)
	}
	if weapon.FireRate != 3.0 {
		t.Errorf("Expected fire rate 3.0, got %f", weapon.FireRate)
	}
	if weapon.MagazineSize != 15 {
		t.Errorf("Expected magazine size 15, got %d", weapon.MagazineSize)
	}
	if weapon.ReloadTime != 1500*time.Millisecond {
		t.Errorf("Expected reload time 1500ms, got %v", weapon.ReloadTime)
	}
	if weapon.ProjectileSpeed != 800.0 {
		t.Errorf("Expected projectile speed 800.0, got %f", weapon.ProjectileSpeed)
	}
	if weapon.Range != 800.0 {
		t.Errorf("Expected range 800.0, got %f", weapon.Range)
	}
	if weapon.Recoil != nil {
		t.Error("Expected nil recoil for Pistol")
	}
}

func TestWeaponConfigToWeapon_WithRecoil(t *testing.T) {
	config := &WeaponConfig{
		Name:              "Uzi",
		Damage:            8,
		FireRate:          10.0,
		MagazineSize:      30,
		ReloadTimeMs:      1500,
		ProjectileSpeed:   800.0,
		Range:             600.0,
		ArcDegrees:        0,
		KnockbackDistance: 0,
		Recoil: &RecoilConfig{
			VerticalPerShot:   2.0,
			HorizontalPerShot: 0.0,
			RecoveryTime:      0.5,
			MaxAccumulation:   20.0,
		},
		SpreadDegrees: 5.0,
	}

	weapon := config.ToWeapon()

	if weapon.Name != "Uzi" {
		t.Errorf("Expected name 'Uzi', got '%s'", weapon.Name)
	}
	if weapon.Recoil == nil {
		t.Fatal("Expected recoil to be non-nil for Uzi")
	}
	if weapon.Recoil.VerticalPerShot != 2.0 {
		t.Errorf("Expected vertical recoil 2.0, got %f", weapon.Recoil.VerticalPerShot)
	}
	if weapon.Recoil.HorizontalPerShot != 0.0 {
		t.Errorf("Expected horizontal recoil 0.0, got %f", weapon.Recoil.HorizontalPerShot)
	}
	if weapon.Recoil.RecoveryTime != 0.5 {
		t.Errorf("Expected recovery time 0.5, got %f", weapon.Recoil.RecoveryTime)
	}
	if weapon.Recoil.MaxAccumulation != 20.0 {
		t.Errorf("Expected max accumulation 20.0, got %f", weapon.Recoil.MaxAccumulation)
	}
	if weapon.SpreadDegrees != 5.0 {
		t.Errorf("Expected spread degrees 5.0, got %f", weapon.SpreadDegrees)
	}
}

func TestWeaponConfigToWeapon_MeleeWeapon(t *testing.T) {
	config := &WeaponConfig{
		Name:              "Bat",
		Damage:            25,
		FireRate:          2.0,
		MagazineSize:      0,
		ReloadTimeMs:      0,
		ProjectileSpeed:   0,
		Range:             90.0,
		ArcDegrees:        80,
		KnockbackDistance: 40,
		Recoil:            nil,
		SpreadDegrees:     0,
	}

	weapon := config.ToWeapon()

	if weapon.Name != "Bat" {
		t.Errorf("Expected name 'Bat', got '%s'", weapon.Name)
	}
	if weapon.MagazineSize != 0 {
		t.Errorf("Expected magazine size 0 for melee, got %d", weapon.MagazineSize)
	}
	if weapon.ReloadTime != 0 {
		t.Errorf("Expected reload time 0 for melee, got %v", weapon.ReloadTime)
	}
	if weapon.Range != 90.0 {
		t.Errorf("Expected range 90.0, got %f", weapon.Range)
	}
	if weapon.ArcDegrees != 80 {
		t.Errorf("Expected arc degrees 80, got %f", weapon.ArcDegrees)
	}
	if weapon.KnockbackDistance != 40 {
		t.Errorf("Expected knockback distance 40, got %f", weapon.KnockbackDistance)
	}
	if !weapon.IsMelee() {
		t.Error("Expected weapon to be identified as melee")
	}
}

func TestGetDefaultConfigPath(t *testing.T) {
	path := GetDefaultConfigPath()
	if path == "" {
		t.Error("Expected non-empty default config path")
	}
	if filepath.Base(path) != "weapon-configs.json" {
		t.Errorf("Expected filename 'weapon-configs.json', got '%s'", filepath.Base(path))
	}
}

func TestValidateWeaponConfig_Valid(t *testing.T) {
	config := &WeaponConfig{
		Name:              "Pistol",
		Damage:            25,
		FireRate:          3.0,
		MagazineSize:      15,
		ReloadTimeMs:      1500,
		ProjectileSpeed:   800.0,
		Range:             800.0,
		ArcDegrees:        0,
		KnockbackDistance: 0,
		Recoil:            nil,
		SpreadDegrees:     0,
		Visuals: WeaponVisuals{
			MuzzleFlashColor:    "0xffdd00",
			MuzzleFlashSize:     8,
			MuzzleFlashDuration: 50,
		},
	}

	err := ValidateWeaponConfig(config)
	if err != nil {
		t.Errorf("Expected valid config to pass validation, got error: %v", err)
	}
}

func TestValidateWeaponConfig_MissingName(t *testing.T) {
	config := &WeaponConfig{
		Name:            "",
		Damage:          25,
		FireRate:        3.0,
		MagazineSize:    15,
		ProjectileSpeed: 800.0,
		Range:           800.0,
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for missing name, got nil")
	}
}

func TestValidateWeaponConfig_InvalidDamage(t *testing.T) {
	config := &WeaponConfig{
		Name:            "Invalid",
		Damage:          0,
		FireRate:        3.0,
		MagazineSize:    15,
		ProjectileSpeed: 800.0,
		Range:           800.0,
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for invalid damage, got nil")
	}
}

func TestValidateWeaponConfig_InvalidFireRate(t *testing.T) {
	config := &WeaponConfig{
		Name:            "Invalid",
		Damage:          25,
		FireRate:        0,
		MagazineSize:    15,
		ProjectileSpeed: 800.0,
		Range:           800.0,
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for invalid fire rate, got nil")
	}
}

func TestValidateWeaponConfig_InvalidRange(t *testing.T) {
	config := &WeaponConfig{
		Name:            "Invalid",
		Damage:          25,
		FireRate:        3.0,
		MagazineSize:    15,
		ProjectileSpeed: 800.0,
		Range:           0,
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for invalid range, got nil")
	}
}

func TestValidateWeaponConfig_RangedWeaponRequiresProjectileSpeed(t *testing.T) {
	config := &WeaponConfig{
		Name:            "InvalidGun",
		Damage:          25,
		FireRate:        3.0,
		MagazineSize:    15, // Has magazine but no projectile speed
		ProjectileSpeed: 0,
		Range:           800.0,
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for ranged weapon without projectile speed, got nil")
	}
}

func TestValidateWeaponConfig_InvalidRecoilRecoveryTime(t *testing.T) {
	config := &WeaponConfig{
		Name:            "InvalidRecoil",
		Damage:          25,
		FireRate:        3.0,
		MagazineSize:    15,
		ProjectileSpeed: 800.0,
		Range:           800.0,
		Recoil: &RecoilConfig{
			VerticalPerShot:   2.0,
			HorizontalPerShot: 1.0,
			RecoveryTime:      0, // Invalid
			MaxAccumulation:   20.0,
		},
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for invalid recoil recovery time, got nil")
	}
}

func TestValidateWeaponConfig_InvalidRecoilMaxAccumulation(t *testing.T) {
	config := &WeaponConfig{
		Name:            "InvalidRecoil",
		Damage:          25,
		FireRate:        3.0,
		MagazineSize:    15,
		ProjectileSpeed: 800.0,
		Range:           800.0,
		Recoil: &RecoilConfig{
			VerticalPerShot:   2.0,
			HorizontalPerShot: 1.0,
			RecoveryTime:      0.5,
			MaxAccumulation:   0, // Invalid
		},
	}

	err := ValidateWeaponConfig(config)
	if err == nil {
		t.Error("Expected error for invalid recoil max accumulation, got nil")
	}
}

func TestLoadWeaponConfigsOrDefault_Success(t *testing.T) {
	// Create temporary config file
	tmpDir := t.TempDir()
	configPath := filepath.Join(tmpDir, "weapon-configs.json")

	config := WeaponConfigFile{
		Version: "1.0.0",
		Weapons: map[string]WeaponConfig{
			"TestWeapon": {
				Name:            "TestWeapon",
				Damage:          50,
				FireRate:        5.0,
				MagazineSize:    20,
				ReloadTimeMs:    2000,
				ProjectileSpeed: 900.0,
				Range:           700.0,
			},
		},
	}

	data, _ := json.MarshalIndent(config, "", "  ")
	os.WriteFile(configPath, data, 0644)

	configs := LoadWeaponConfigsOrDefault(configPath)
	if configs == nil {
		t.Fatal("Expected configs to be non-nil")
	}
	if configs["TestWeapon"] == nil {
		t.Error("Expected TestWeapon to be loaded")
	}
}

func TestLoadWeaponConfigsOrDefault_FallbackToHardcoded(t *testing.T) {
	// Test with non-existent path
	configs := LoadWeaponConfigsOrDefault("/nonexistent/path.json")
	if configs == nil {
		t.Fatal("Expected configs to be non-nil (fallback)")
	}

	// Should have hardcoded weapons
	expectedWeapons := []string{"Pistol", "Bat", "Katana", "Uzi", "AK47", "Shotgun"}
	for _, weaponName := range expectedWeapons {
		if configs[weaponName] == nil {
			t.Errorf("Expected hardcoded weapon '%s' to exist in fallback", weaponName)
		}
	}
}
